### PR TITLE
hive/metrics: Adds hive degraded status labeled by modules

### LINKED ISSUE #34823: hive: Add new metric to expose degraded hive modules
Currently, we expose the `cilium_hive_status` but we have no insights on the actual affected modules and we need to manually investigate which modules are affected.

Add a new metric to only highlight the degraded status labeled by the module name. 

Considered concerns: 
* High metric cardinality -> the metric only publish data for degraded modules, the amount of degraded modules should be minimal and expected the metric to have no data in a healthy cluster.

### PR DESCRIPTION
**Pull request by ovidiutirla**
Adds `cilium_hive_degraded_status` metric to count degraded health status levels of Hive components labeled by modules. It improves the monitoring effectiveness by exposing the affected modules instead of the general count.

Fixes: https://github.com/cilium/cilium/issues/34823

```release-note
Adds `cilium_hive_degraded_status` metric to count degraded health status levels of Hive components labeled by modules. ```

### CODE REVIEW

**Comment by ovidiutirla** [comment]
<details>
  <summary>Grafana</summary>
If there are no cells degraded, the metric will have no data to reduce the metrics cardinality.

image

image

</details>

**Comment by ovidiutirla** [comment]
Please review but do not merge it yet as we need to wait for the hive version bump to ensure the job modules IDs are complaint.

File: pkg/hive/health/metrics.go
Comment (bimmlerd) [inline pkg/hive/health/metrics.go:35]: I'm not a metrics expert but doesn't this potentially have issues with cardinality? cc @cilium/metrics

**Review by bimmlerd — APPROVED** [review]
foundations-wise this is fine, but requesting a review from metrics.

File: pkg/hive/health/metrics.go
Comment (ovidiutirla) [reply pkg/hive/health/metrics.go:35]: Hey @bimmlerd, the cardinality is not expected to be a problem as we are expecting to have a small (or none) number of cells health to be degraded. If there are no cells degraded, the metric will have no data to reduce the metrics cardinality. We are only publishing module names for the degraded cells. Thanks for the review.

Comment (learnitall) [reply pkg/hive/health/metrics.go:35]: It looks like the value of this metric can only go up, since it's never decremented. Can we change it to a counter?

**Review by learnitall — COMMENTED** [review]
@tommyp1ckles is the expert, so please take the following with a grain of salt. I'm just giving a drive-by review :).

If all modules within the hive are defined when cilium boots up, meaning, no modules can be added during runtime, then I think we will be fine. The biggest concern would be a metric with unbounded labels, like "pod name", but since the number of labels here is determined when the agent boots we won't have to worry about unbounded growth.

That being said, one concern I do have is that the optimization "We are only publishing module names for the degraded cells" isn't entirely true. When a module experiences degraded health, the metric with the module's label will be exposed until the agent is restarted even if the module becomes healthy again. As soon as `WithLabelValues` is called, a metric tracking the given label is exposed. Even if the PR was modified to decrement the module's metric value when it becomes healthy again, the metric's value will be `0` and therefore continue to be exposed.

The impact of the metric's cardinality will change as new cells are added and cell behavior changes, which is also a concern because that kind of impact is not easy to review or detect.

**Comment by ovidiutirla** [comment]
> @tommyp1ckles is the expert, so please take the following with a grain of salt. I'm just giving a drive-by review :).
> 
> If all modules within the hive are defined when cilium boots up, meaning, no modules can be added during runtime, then I think we will be fine. The biggest concern would be a metric with unbounded labels, like "pod name", but since the number of labels here is determined when the agent boots we won't have to worry about unbounded growth.
> 
> That being said, one concern I do have is that the optimization "We are only publishing module names for the degraded cells" isn't entirely true. When a module experiences degraded health, the metric with the module's label will be exposed until the agent is restarted even if the module becomes healthy again. As soon as `WithLabelValues` is called, a metric tracking the given label is exposed. Even if the PR was modified to decrement the module's metric value when it becomes healthy again, the metric's value will be `0` and therefore continue to be exposed.
> 
> The impact of the metric's cardinality will change as new cells are added and cell behavior changes, which is also a concern because that kind of impact is not easy to review or detect.

I see some benefits a metric like this one, that reports the cell modules that are in a degraded state, do you see any value in having something like this?

To reduce the risk of cardinality. do you think it would make sense to limit the number of reported modules, for example if we exceed to have more than 10 unhealthy modules, we publish the metrics only for the first 10?

**Comment by learnitall** [comment]
> I see some benefits in a metric like this one, that reports the cell modules that are in a degraded state, do you see any value in having something like this?

I agree there's definitely a benefit to having it, we just have to figure out a "safe" way to add it into the agent to prevent potential issues.

> To reduce the risk of cardinality. do you think it would make sense to limit the number of reported modules, for example if we exceed to have more than 10 unhealthy modules, we publish the metrics only for the first 10?

This could be interesting, but the main concern I'd have with an approach like this is with agents that run for a very long time. If an agent faces an issue that causes some modules to fail, say in the first day or two that it is running, then the metric is only useful if those same modules are the only ones to fail for the rest of the lifetime of the agent. We won't have visibility into other failing modules.

One idea that @marseel, @joamaki and I were discussing is having a second Prometheus endpoint for high-cardinality metrics that are useful for troubleshooting, like this one. This shifts them to be more "opt-in" for continual scraping and opens the option for them to only be used when needed to debug issues. Would you have any interest in implementing something like this for Cilium?

**Comment by ovidiutirla** [comment]
> > I see some benefits in a metric like this one, that reports the cell modules that are in a degraded state, do you see any value in having something like this?
> 
> I agree there's definitely a benefit to having it, we just have to figure out a "safe" way to add it into the agent to prevent potential issues.
> 
> > To reduce the risk of cardinality. do you think it would make sense to limit the number of reported modules, for example if we exceed to have more than 10 unhealthy modules, we publish the metrics only for the first 10?
> 
> This could be interesting, but the main concern I'd have with an approach like this is with agents that run for a very long time. If an agent faces an issue that causes some modules to fail, say in the first day or two that it is running, then the metric is only useful if those same modules are the only ones to fail for the rest of the lifetime of the agent. We won't have visibility into other failing modules.
> 
> One idea that @marseel, @joamaki and I were discussing is having a second Prometheus endpoint for high-cardinality metrics that are useful for troubleshooting, like this one. This shifts them to be more "opt-in" for continual scraping and opens the option for them to only be used when needed to debug issues. Would you have any interest in implementing something like this for Cilium?

So thinking more about it, how many cells could we expect to be degraded and the cluster to be happy...? 
And how many cell.modules do we have, 50? So even if we have all cells degraded this is still a low cardinality... 
We are exposing the cell.moduleID and not all the sub-modules. 


I think the new API would make sense and I'm happy to contribute to it, but for this one I think it would make sense to have it enabled by default on the default API.

**Comment by learnitall** [comment]
> I think the new API would make sense and I'm happy to contribute to it, but for this one I think it would make sense to have it enabled by default on the default API.

Thank you, that's kind of you. I think that your logic makes sense here. IMO we can move forward with this PR then. CC @tommyp1ckles can you take a look?

File: pkg/hive/health/metrics.go
Comment (ovidiutirla) [reply pkg/hive/health/metrics.go:35]: It could get decremented when the cells are able to reconcile in a way they become healthy again.

Comment (learnitall) [reply pkg/hive/health/metrics.go:35]: That makes sense and would be really cool, but I don't see the logic in the code to decrement the gauge when the modules become healthy again. Could we add it?

Comment (ovidiutirla) [reply pkg/hive/health/metrics.go:35]: They should automatically get decremented on each iteration when we fetch the modules status.

Comment (learnitall) [reply pkg/hive/health/metrics.go:35]: Ah ok it clicked, sorry about that!

Comment (tommyp1ckles) [reply pkg/hive/health/metrics.go:35]: I do like that you've limited this to just the module, instead of the full path.  So the cardinality of this metric should be fixed, at least within a release/configuration.

However I am still a bit concerned about the overall cardinality here for a metric that's enabled by default (maybe this one should be opt-in with the IsDisabled flag option?).

> If there are no cells degraded, the metric will have no data to reduce the metrics cardinality. We are only publishing module names for the degraded cells.

In this case, would it make sense to make this a "Deletable" Gauge Vec type and remove modules that are no longer degraded? Otherwise when a module gets marked degraded it will stay that way for the lifetime of the agent.

**Comment by tommyp1ckles** [comment]
@learnitall @ovidiutirla Sorry for the late review, I'm a bit on the fence on this one. Theoretically it's unbounded cardinality but in practice there is only so many modules in a particular release/configuration that could be emitted.

One suggestion I had was to use the DeletableGaugeVec type to manage the metric lifecycles, such that we cleanup modules that are zero instead of keeping them around forever. 

Otherwise, just making this metric Opt-in via the IsDisabled option in the metrics-opts would remove any concern 😄

**Comment by ovidiutirla** [comment]
@tommyp1ckles, kindly ask for a review. I applied your suggestions, marked as Deletable and also disabled by default 🙏

File: pkg/hive/health/metrics.go
Comment (tommyp1ckles) [inline pkg/hive/health/metrics.go:68]: One bit of an edge case with this is if all health statuses for with the same moduleID get deleted then it won't get cleaned up here as presumably it won't even show up as a 0 in the map.  

However I don't think this ever actually happens - given this metric isn't on by default, it's probably not worth the extra complexity of making this this will always clean up every module key.

**Review by tommyp1ckles — APPROVED** [review]

File: pkg/hive/health/metrics.go
Comment (joamaki) [inline pkg/hive/health/metrics.go:107]: How does this work with duplicate module ids? Say we have two modules ”a.c.foo.reconciler” and ”a.c.bar.reconciler” and one is healthy and one is degraded. How would that show up?

Comment (ovidiutirla) [reply pkg/hive/health/metrics.go:107]: `ID.Module` is actually the  `FullModuleID is the fully qualified module identifier, e.g. the concat of nested module ids, e.g. "agent.controlplane.endpoint-manager". Provided in the module's scope.`

So we should mark as degraded, `a.c.foo` or `a.c.bar`. We are marking as degraded module not the components if that makes sense?

Take the following example:

`ID.Module.String()`: `agent.controlplane.cilium-agent-dynamic-config`
`ID.String()`: `agent.controlplane.cilium-agent-dynamic-config.job-k8s-reflector-cilium-configs-cm-cilium-config-kube-system`

File: pkg/hive/health/metrics.go
Comment (ovidiutirla) [reply pkg/hive/health/metrics.go:68]:
```
                         _, ok := idToStatus[obj.ID.Module.String()]
			if obj.Level == types.LevelDegraded {
				idToStatus[obj.ID.Module.String()]++
			} else if !ok {
				idToStatus[obj.ID.Module.String()] = 0
			}
```

All module ids should show up in the map as 0? I'm not quite following, could you please clarify it a bit?

Comment (tommyp1ckles) [reply pkg/hive/health/metrics.go:68]: For reference, I discussed this with @ovidiutirla on slack, the specific condition I was getting at here is:

Say we just had one reporter: x.a -> degraded.

* metrics has module 'x' = 1
* x.a tree is removed via [Close](https://github.com/cilium/cilium/blob/724d5b5f3092a0f329d9b36e674adfa020c27188/pkg/hive/health/provider.go#L228-L232).
* x.a is removed [from statedb table](https://github.com/cilium/cilium/blob/724d5b5f3092a0f329d9b36e674adfa020c27188/pkg/hive/health/provider.go#L112)
This triggers the statedb metrics watch channel, causing metrics iteration.
* idToStatus is counted up, but it doesn't contain module x.
* We iterate all module counts in idToStatus, but we fail to remove the metric row... so the metric is left at 1 indefinitely.

File: pkg/hive/health/metrics.go
Comment (tommyp1ckles) [reply pkg/hive/health/metrics.go:107]: @joamaki broadly speaking, this metric is a count of degraded health reports with the same moduleID prefix.

File: pkg/hive/health/metrics.go
Comment (tommyp1ckles) [reply pkg/hive/health/metrics.go:68]: > One bit of an edge case with this is if all health statuses for with the same moduleID get deleted then it won't get cleaned up here as presumably it won't even show up as a 0 in the map.

Thinking through this a bit more, this might be more likely than I thought, since the[ Close function deletes all sub-component reports in the module](https://github.com/cilium/cilium/blob/724d5b5f3092a0f329d9b36e674adfa020c27188/pkg/hive/health/provider.go#L109-L118), all it would take is a Close() on a root reporter to cause a leak.

Comment (ovidiutirla) [reply pkg/hive/health/metrics.go:68]: @tommyp1ckles, thanks for the discussion we had on Slack. I've updated the PR to also handle your concern. Please have another look before merging this.

**Comment by ovidiutirla** [comment]
cc @pchaigno :)

**Pull request merged**
